launch/launch/descriptions/executable.py

```diff
@@ -0,0 +1,141 @@
+            arguments to the executable, each item may be a string or a list of strings
+            and Substitutions to be resolved at runtime
+        :param name: The label used to represent the process, as a string or a Substitution
+            to be resolved at runtime, defaults to the basename of the executable
+        :param cwd: The directory in which to run the executable
+        :param env: Dictionary of environment variables to be used, starting from a clean
+            environment. If None, the current environment is used.
+        :param additional_env: Dictionary of environment variables to be added. If env was
+            None, they are added to the current environment. If not, env is updated with
+            additional_env.
+        """
+        self.__cmd = [normalize_to_list_of_substitutions(x) for x in cmd]
+        self.__name = name if name is None else normalize_to_list_of_substitutions(name)
+        self.__cwd = cwd if cwd is None else normalize_to_list_of_substitutions(cwd)
+        self.__env = None  # type: Optional[List[Tuple[List[Substitution], List[Substitution]]]]
+        if env is not None:
+            self.__env = []
+            for key, value in env.items():
+                self.__env.append((
+                    normalize_to_list_of_substitutions(key),
+                    normalize_to_list_of_substitutions(value)))
+        self.__additional_env: Optional[List[Tuple[List[Substitution], List[Substitution]]]] = None
+        if additional_env is not None:
+            self.__additional_env = []
+            for key, value in additional_env.items():
+                self.__additional_env.append((
+                    normalize_to_list_of_substitutions(key),
+                    normalize_to_list_of_substitutions(value)))
+
+    @property
+    def name(self):
+        """Getter for name."""
+        return self.__name
+
+    @property
+    def cmd(self):
+        """Getter for cmd."""
+        return self.__cmd
+
+    @property
+    def cwd(self):
+        """Getter for cwd."""
+        return self.__cwd
+
+    @property
+    def env(self):
+        """Getter for env."""
+        return self.__env
+
+    @property
+    def additional_env(self):
+        """Getter for additional_env."""
+        return self.__additional_env
+
+    @property
+    def process_details(self):
+        """Getter for the substituted executable details, e.g. cmd, cwd, env, or None if substitutions have not been performed."""
+        return self.__process_event_args
+
+    def apply_context(self, context: LaunchContext):
+        """
+        Prepares an executable description for execution in a given environment.
+
+        This does the following:
+        - performs substitutions on various properties
+        """
+        self.__expand_substitutions(context)
+        process_event_args = self.__process_event_args
+        if process_event_args is None:
```

Comment:
nit: this can be done in a single line
```suggestion
        if self.__process_event_args is None:
```

Reply:
is this check even needed?

Reply:
Removed the identified check. `__expand_substitutions` is still modifying the object, but it's doing is by adding newly-defined values instead of changing the existing ones. Not sure if it's really different enough for this concern, so if we want to continue tweaking it, please let me know. If we really would rather have a stateful object to pass around instead of changing this one, I think I can see the outlines of doing that, but I'm not sure if we're gaining much (other than complexity). Open to suggestions/thoughts/comments, of course.